libraries/ms-common/src/stm32f0xx/adc.c

```diff
@@ -0,0 +1,159 @@
+
+typedef struct ADCInterrupt {
+  ADCCallback callback;
+  void *context;
+  uint16_t reading;
+} ADCInterrupt;
+
+typedef struct ADCStatus {
+  ADCChannel current_channel;
+  bool continuous;
+} ADCStatus;
+
+static ADCInterrupt s_adc_interrupts[NUM_ADC_CHANNEL];
+static ADCStatus s_adc_status;
+
+static prv_get_ffs(uint32_t select, ADCChannel adc_channel) {
+  select = select >> adc_channel + 1;
+  uint8_t offset = 0;
+
+  uint32_t ffs_max = NUM_ADC_CHANNEL - adc_channel;
+  for (uint32_t i = 0; i < ffs_max; i++) {
+    if (select & (1 << i)) {
+      return (i + adc_channel + 1);
+    }
+  }
+  return 0;
+}
+
+void adc_init(ADCMode adc_mode) {
+  // Stop ongoing conversions and disable the ADC 
+  ADC_StopOfConversion(ADC1);
+  while (ADC_GetFlagStatus(ADC1, ADC_FLAG_ADSTP)) { }
+  
+  ADC1->CR |= ADC_FLAG_ADDIS;
+  while (ADC_GetFlagStatus(ADC1, ADC_FLAG_ADEN)) { }
+
+  // Once the ADC has been reset, enable it with the given settings
+  RCC_APB2PeriphClockCmd(RCC_APB2Periph_ADC1, ENABLE);
+
+  ADC_InitTypeDef adc_settings = {
+    .ADC_Resolution = ADC_Resolution_12b,
+    .ADC_ContinuousConvMode = adc_mode,
+    .ADC_ExternalTrigConvEdge = ADC_ExternalTrigConvEdge_None,
+    .ADC_ExternalTrigConv = ADC_ExternalTrigConv_T1_TRGO,
+    .ADC_DataAlign = ADC_DataAlign_Right,
+    .ADC_ScanDirection = ADC_ScanDirection_Upward,
+  };
+
+  ADC_Init(ADC1, &adc_settings);
+
+  // Calculate the ADC calibration factor
+  ADC_GetCalibrationFactor(ADC1);
+
+  ADC_ContinuousModeCmd(ADC1, adc_mode);
+  while (ADC_GetFlagStatus(ADC1, ADC_FLAG_ADCAL)) { }
+
+  // Enable the ADC
+  ADC_Cmd(ADC1, ENABLE);
+  while (!ADC_GetFlagStatus(ADC1, ADC_FLAG_ADRDY)) { }
+
+  // Set the ADC to wait mode
+  ADC_WaitModeCmd(ADC1, ENABLE);
+  ADC_AutoPowerOffCmd(ADC1, ENABLE);
+
+  // Enable interrupts for the end of each conversion
+  stm32f0xx_interrupt_nvic_enable(ADC1_COMP_IRQn, INTERRUPT_PRIORITY_NORMAL);
+  ADC_ITConfig(ADC1, ADC_IER_EOCIE, ENABLE);
+
+  // Initialize static varables
+  s_adc_status.continuous = adc_mode;
+  s_adc_status.current_channel = 0;
+
+  // Start background conversions if in continuous
+  if (adc_mode) {
+    ADC_StartOfConversion(ADC1);
+  }
+}
+
+StatusCode adc_set_channel(ADCChannel adc_channel, bool new_state) {
+  if (adc_channel >= NUM_ADC_CHANNEL) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+
+  if (new_state) {
+    ADC_ChannelConfig(ADC1, (1 << adc_channel), ADC_SampleTime_239_5Cycles);
+  } else {
+    ADC1->CHSELR &= !(1 << adc_channel);
+  }
+
+  return STATUS_CODE_OK;
+}
+
+StatusCode adc_register_callback(ADCChannel adc_channel, ADCCallback callback, void *context) {
+  // Returns invalid if the given address is not connected to an ADC channel
+  if (adc_channel >= NUM_ADC_CHANNEL) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+
+  s_adc_interrupts[adc_channel].callback = callback;
+  s_adc_interrupts[adc_channel].context = context;
+
+  return STATUS_CODE_OK;
+}
+
+uint16_t adc_read_value(ADCChannel adc_channel) {
```

Comment:
You don't have any error checking on this function, so it's possible for the ADC channel to be invalid.